src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py

```diff
@@ -0,0 +1,442 @@
+
+        cached_student_layers: Dict[str, torch.nn.Module] = {}
+        if self._student_layer_names is None:
+            _find_layers_by_type(module, cached_student_layers)
+            self._student_layer_names = list(cached_student_layers.keys())
+        else:
+            _find_layers_by_name(
+                module, self._student_layer_names, cached_student_layers
+            )
+        _LOGGER.info("Distilling student layers: %s", self._student_layer_names)
+
+        cached_teacher_layers: Dict[str, torch.nn.Module] = {}
+        if self._teacher_layer_names is None:
+            _find_layers_by_type(self._teacher, cached_teacher_layers)
+            self._teacher_layer_names = list(cached_teacher_layers.keys())
+        else:
+            _find_layers_by_name(
+                self._teacher, self._teacher_layer_names, cached_teacher_layers
+            )
+        _LOGGER.info("Distilling teacher layers: %s", self._teacher_layer_names)
+
+        if len(self._teacher_layer_names) != len(self._student_layer_names):
+            raise ValueError(
+                "Found different numbers of teacher and student layers to distill. "
+                "Set teacher_layer_names and student_layer_names explicitly."
+            )
+
+        self._student_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_student_output, self._student_output_shapes
+                )
+            )
+            for name, layer in cached_student_layers.items()
+        ]
+
+        self._teacher_handles = [
+            layer.register_forward_hook(
+                _create_cache_output_hook(
+                    name, self._cached_teacher_output, self._teacher_output_shapes
+                )
+            )
+            for name, layer in cached_teacher_layers.items()
+        ]
+
+    def update(
+        self,
+        module: Module,
+        optimizer: torch.optim.Optimizer,
+        epoch: float,
+        steps_per_epoch: int,
+    ):
+        super().update(module, optimizer, epoch, steps_per_epoch)
+        if epoch >= self.end_epoch:
+            for handle in self._student_handles:
+                handle.remove()
+            for handle in self._teacher_handles:
+                handle.remove()
+            self._reset_cache()
+
+    def compute_total_loss(self, loss, distillation_loss):
+        return loss + self.gain * distillation_loss
+
+    def compute_distillation_loss(self, optimizer: torch.optim.Optimizer, **kwargs):
+        if self.project_features and len(self._projection) == 0:
+            # NOTE: have to call initialize here because we need the cached output
+            # from the module. i.e. we need forward to have been called already
+            projection_params = self._initialize_projection()
+            optimizer.add_param_group({"params": projection_params})
+
+            # NOTE: since we added a param group, if someone tries
```

Comment:
let's add support for re-serialization, so information is not lost.

a few notes

`state_dict()`
1. we'll serialize everything to the **modifier** state dict
2. we'll include all the projection weights (projection module name -> `projection.state_dict()`
3. we'll include the optimizer param group dict

`load_state_dict()`
1. if projections have not been created yet (this is the likely case) - we should just save the info as a private property of the modifier and then load in the weights for the projections/optimizer as soon as the projections are added
2. if the projections have been created, we can load directly in

@corey-nm @rahul-tuli  let me know if this is clear. can go over offline if needed

Reply:
@bfineran kindly check, the new implementation is up